File: Algo Problems/Backtracking/Sudoku/sudoku.py
def isSafe(board, x, y, val):
    for m in range(len(board[0])):
        if val == board[m][y]: return False

    for n in range(len(board[0])):
        if val == board[x][n]: return False

    smi = (x//3) * 3
    smj = (y//3) * 3

    for m in range(3):
        for n in range(3):
            if board[smi+m][smj+n] == val:
                return False
    
    return True



def solve(board, i, j):

    if i == len(board):
        print(board)
        return 
    
    ni = 0
    nj = 0
    if j == len(board[0])-1: 
        ni = i+1
        nj = 0
    else: 
        ni = i
        nj = j+1

    if board[i][j] != 0:
        solve(board, ni, nj)
    else:
        for po in range(1,10):
            if isSafe(board, i, j, po):
                board[i][j] = po
                solve(board, ni, nj)
                board[i][j] = 0


sudoku = [[0,0,1,0,3,9,0,0,0],
          [0,7,2,0,0,0,0,9,0],
          [0,0,0,0,0,6,0,8,0],
          [0,5,0,0,0,0,0,0,0],
          [0,0,0,2,0,0,0,5,0],
          [3,0,0,0,0,0,0,0,0],
          [0,0,0,0,1,0,3,0,0],
          [0,4,0,0,0,0,0,0,0],
          [9,0,0,0,4,0,0,0,5]
        ]


solve(sudoku, 0, 0)

File: Algo Problems/Backtracking/Playlist Problems/Chess Type/sudoku.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
    private: 
        vector<vector<int>> board;
        int size;
        bool solved = false;
    public: 
        Solution(vector<vector<int>> b){
            board = b;
            size = 9;
        }

        void displayBoard(){
            cout<<"Board ===================\n"<<endl;
            for(auto it: board){
                for(auto x: it) {
                    cout<<x<<" ";
                }
                cout<<endl;
            }
            cout<<"===================\n"<<endl;
        }

        bool isSafe(int row, int col, int num){
            for(int i=0;i<size;i++){
                if(board[row][i]==num) return false;
                if(board[i][col]==num) return false;
            }
            int rs=(row/3)*3, cs=(col/3)*3;

            for(int i=0;i<3;i++){
                for(int j=0;j<3;j++){
                    if(board[i+rs][j+cs]==num) return false;
                }
            }
            return true;
        }

        void util(int row, int col){
            if(solved) return;
            if(row == size){
                displayBoard();
                solved = true;
                return;
            }
            int nr = 0, nc = 0;
            if(col == size-1) {
                nr = row+1;
                nc = 0;
            } else{
                nr = row;
                nc = col+1;
            }

            if(board[row][col]!=0) util(nr, nc);
            else{
                for(int n=1;n<=9;n++){
                    if(isSafe(row, col, n)){
                        board[row][col] = n;
                        util(nr, nc);
                        board[row][col] = 0;
                    }
                }
            }
        }

        void solve(){
            util(0, 0);
        }
};

int main(){
    Solution s({
        {0,0,1,0,3,9,0,0,0},
        {0,7,2,0,0,0,0,9,0},
        {0,0,0,0,0,6,0,8,0},
        {0,5,0,0,0,0,0,0,0},
        {0,0,0,2,0,0,0,5,0},
        {3,0,0,0,0,0,0,0,0},
        {0,0,0,0,1,0,3,0,0},
        {0,4,0,0,0,0,0,0,0},
        {9,0,0,0,4,0,0,0,5}
    });
    cout<<"Hello World"<<endl;
    s.solve();
}